lib/std/crypto/25519/curve25519.zig

```diff
@@ -0,0 +1,154 @@
+const std = @import("std");
+
+/// Group operations over Curve25519.
+pub const Curve25519 = struct {
+    /// The underlying prime field.
+    pub const Fe = @import("field.zig").Fe;
+    /// Field arithmetic mod the order of the main subgroup.
+    pub const scalar = @import("scalar.zig");
+
+    x: Fe,
+
+    /// Decode a Curve25519 point from its compressed (X) coordinates.
+    pub inline fn fromBytes(s: [32]u8) Curve25519 {
+        return .{ .x = Fe.fromBytes(s) };
+    }
+
+    /// Encode a Curve25519 point.
+    pub inline fn toBytes(p: Curve25519) [32]u8 {
+        return p.x.toBytes();
+    }
+
+    /// Return the Curve25519 base point.
+    pub inline fn basePoint() Curve25519 {
+        return .{ .x = Fe.curve25519BasePoint };
+    }
+
+    /// Check that the encoding of a Curve25519 point is canonical.
+    pub fn rejectNonCanonical(s: [32]u8) !void {
+        return Fe.rejectNonCanonical(s, false);
+    }
+
+    /// Reject the neutral element.
+    pub fn rejectIdentity(p: Curve25519) !void {
+        if (p.x.isZero()) {
+            return error.IdentityElement;
+        }
+    }
+
+    fn ladder(p: Curve25519, s: [32]u8, comptime bits: usize) !Curve25519 {
+        var x1 = p.x;
+        var x2 = Fe.one;
+        var z2 = Fe.zero;
+        var x3 = x1;
+        var z3 = Fe.one;
+        var swap: u8 = 0;
+        var pos: usize = bits - 1;
+        while (true) {
+            const b = (s[pos / 8] >> @intCast(u3, pos & 7)) & 1;
```

Comment:
Does it produce better code?

It seems to slightly decreases readability. Having `/ 8` and `& 7` next to each other immediately captures the intent, while I really need to read the truncation twice to understand what it does.

Reply:
`@intCast` should be the least preferable cast: every time you see `@intCast` in some code it should be a read-it-3-times to be sure situation; `@as` and `@truncate` are meant to be understandable at a glance.

Perhaps to make things a bit more symmetrical, use `>> 3` instead of `/ 8`? which would pair nicely with the `u3`.

Reply:
Done.